java/fury-core/src/main/java/org/apache/fury/serializer/JdkProxySerializer.java

```diff
@@ -20,16 +20,30 @@
 package org.apache.fury.serializer;
 
 import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
 import java.lang.reflect.Proxy;
 import org.apache.fury.Fury;
 import org.apache.fury.memory.MemoryBuffer;
+import org.apache.fury.resolver.RefResolver;
+import org.apache.fury.util.Platform;
 import org.apache.fury.util.Preconditions;
 import org.apache.fury.util.ReflectionUtils;
 
 /** Serializer for jdk {@link Proxy}. */
 @SuppressWarnings({"rawtypes", "unchecked"})
 public class JdkProxySerializer extends Serializer {
 
+  // Make offset compatible with graalvm native image.
+  private static final long PROXY_HANDLER_FIELD_OFFSET =
```

Comment:
The try catch block is still needed, graalvm only capture this pattern and replace the offset. This invocation can be placed in catch block

Reply:
Should I leave the InvocationHandler stub initialization as is or move it to the static block as well? I'm new to grallvm but it's on my todo list 👍

Reply:
Should I leave the InvocationHandler stub initialization as is or move it to the static block as well? I'm new to grallvm but it's on my todo list 👍

Reply:
Leave it as it looks good to me

Reply:
done 👍

Reply:
Okay, I just stumbled over the try...catch mention. Is a try / catch really necessary or is the static block okay? 🤔

Reply:
Some jdk may have different field name for `InvocationHandler` instead of using `h`

Reply:
I'm asking because this new [Platform.objectFieldOffset(clazz, fieldType)](https://github.com/apache/incubator-fury/pull/1365/files#diff-7675c883d6e1c72d376ffb7941552f0b28f6821b491bce97d2838bce9bd8992cR155) variant doesn't throw a `NoSuchFieldException`

Reply:
Or we can leave it in another PR.

If we implement in this PR, we may need an another inner class to hold offset like:
```java
 private static class Offset {
    private static final long ARRAY_LIST_SIZE_OFFSET;
    private static final long ARRAY_LIST_ARRAY_OFFSET;

    static {
      try {
        ARRAY_LIST_SIZE_OFFSET =
            Platform.objectFieldOffset(ArrayList.class.getDeclaredField("size"));
        ARRAY_LIST_ARRAY_OFFSET =
            Platform.objectFieldOffset(ArrayList.class.getDeclaredField("elementData"));
      } catch (NoSuchFieldException e) {
        throw new RuntimeException(e);
      }
    }
  }
```

Then use that offset in `org.apache.fury.serializer.JdkProxySerializer#read` with a try catch block.

Reply:
> I'm asking because this new [Platform.objectFieldOffset(clazz, fieldType)](https://github.com/apache/incubator-fury/pull/1365/files#diff-7675c883d6e1c72d376ffb7941552f0b28f6821b491bce97d2838bce9bd8992cR155) variant doesn't throw a `NoSuchFieldException`

This pattern can be captured at graalvm native image build time. It should be used only in fallback block.

Reply:
see more in:
-  https://developers.redhat.com/articles/2022/05/09/using-unsafe-safely-graalvm-native-image#unsafe_proves_to_be_unsafe
- https://www.graalvm.org/latest/reference-manual/native-image/dynamic-features/Reflection/#unsafe-accesses

Reply:
Do you have something like this in mind?
```java
static {
    try {
      // Make offset compatible with graalvm native image.
      PROXY_HANDLER_FIELD_OFFSET =
          Platform.objectFieldOffset(Proxy.class.getDeclaredField("h"));
    } catch (NoSuchFieldException e) {
      PROXY_HANDLER_FIELD_OFFSET = Platform.objectFieldOffset(Proxy.class, InvocationHandler.class);
    }
}
```
But I think i have to work with a temp assignment...

Reply:
This compiles but I'm not sure if it works for graalvm?
```java
  static {
    long proxyHanlderFieldOffset;
    try {
      // Make offset compatible with graalvm native image.
      proxyHanlderFieldOffset = Platform.objectFieldOffset(Proxy.class.getDeclaredField("h"));
    } catch (NoSuchFieldException e) {
      proxyHanlderFieldOffset = Platform.objectFieldOffset(Proxy.class, InvocationHandler.class);
    }
    PROXY_HANDLER_FIELD_OFFSET = proxyHanlderFieldOffset;
  }
```

Reply:
Okay, I studied the redhat documentation in the link provided. So it seems for graalvm we assume that the name of the field is alway's 'h' and for other JVM's it could differ.

I updated the PR with this variant:
```java
  static {
    if (GraalvmSupport.isGraalBuildtime()) {
      try {
        // Make offset compatible with graalvm native image.
        PROXY_HANDLER_FIELD_OFFSET = Platform.objectFieldOffset(Proxy.class.getDeclaredField("h"));
      } catch (NoSuchFieldException e) {
        throw new RuntimeException(e);
      }
    } else {
      // not all JVM implementations use 'h' as internal InvocationHandler name
      PROXY_HANDLER_FIELD_OFFSET =
          ReflectionUtils.getFieldOffset(Proxy.class, InvocationHandler.class);
    }
  }
```
My intention is to use the graalvm 'automatic fix' pattern when it's executed during graalvm build time and the other variant when executed during normal JVM execution.
Could this work?

Reply:
I'm not sure, may need a test. Could we add an test in integration_tests/graalvm_tests?
FYI, I tested following pattern, it works too:
```java
  private static final Field field;
  private static final long offset;

  static {
    field = ReflectionUtils.getField(Example.class, "f2");
    offset = Platform.objectFieldOffset(field);
  }
```

Reply:
Proxy in graalvm is a little complex, may need to add a json config file: https://www.graalvm.org/latest/reference-manual/native-image/guides/configure-dynamic-proxies/